### PR TITLE
Clean up firebaseUtils #630

### PR DESCRIPTION
**Pull request by paulohp**
Following the @jamestalmage instructions on #630, I made the clean up. Follow the progress:
- [x] utils.defer ==> $q.defer
- [x] utils.reject ==> $q.reject
- [x] utils.resolve ==> $q.when
- [x] utils.compile ==> $rootScope.$evalAsync
- [ ] utils.batch ==> $rootScope.$evalAsync
- [x] utils.applyDefaults ==> angular.extend 
- [ ] utils.wait - only used once inside of utils.debounce
- [x] utils.dataKeys - never used

### CODE REVIEW

**Comment by paulohp** [comment]
I signed it!

File: src/FirebaseArray.js [hunk 1]
```diff
@@ -314,7 +314,7 @@
             }
             rec.$id = $firebaseUtils.getKey(snap);
             rec.$priority = snap.getPriority();
-            $firebaseUtils.applyDefaults(rec, this.$$defaults);
+            angular.extend(rec, this.$$defaults);
```
Comment (katowulf) [inline src/FirebaseArray.js]: Defaults should only be applied if the key does not exist. This will always overwrite the existing value with the defaults, which isn't desirable.

File: src/FirebaseArray.js [hunk 2]
```diff
@@ -354,7 +354,7 @@
           if( angular.isObject(rec) ) {
             // apply changes to the record
             changed = $firebaseUtils.updateRec(rec, snap);
-            $firebaseUtils.applyDefaults(rec, this.$$defaults);
+            angular.extend(rec, this.$$defaults);
```
Comment (katowulf) [inline src/FirebaseArray.js]: Same here.

File: src/FirebaseObject.js [hunk 3]
```diff
@@ -59,7 +59,7 @@
         this.$id = $firebaseUtils.getKey(ref.ref());
         this.$priority = null;
 
-        $firebaseUtils.applyDefaults(this, this.$$defaults);
+        angular.extend(this, this.$$defaults);
```
Comment (katowulf) [inline src/FirebaseObject.js]: and here

File: src/FirebaseObject.js [hunk 4]
```diff
@@ -203,7 +203,7 @@
           // applies new data to this object
           var changed = $firebaseUtils.updateRec(this, snap);
           // applies any defaults set using $$defaults
-          $firebaseUtils.applyDefaults(this, this.$$defaults);
+          angular.extend(this, this.$$defaults);
```
Comment (katowulf) [inline src/FirebaseObject.js]: and here

File: tests/unit/FirebaseArray.spec.js [hunk 5]
```diff
@@ -570,7 +570,7 @@ describe('$firebaseArray', function () {
 
     it('should apply $$defaults if they exist', function() {
       var arr = stubArray(null, $firebaseArray.$extend({
-        $$defaults: {aString: 'not_applied', foo: 'foo'}
+        $$defaults: {aString: 'alpha', foo: 'foo'}
```
Comment (katowulf) [inline tests/unit/FirebaseArray.spec.js]: Why were these changed? The tests should still be valid as-is if the new code works. Also, "not_applied" describes the desired outcome in this case where "alpha" is non-descript.

File: tests/unit/FirebaseArray.spec.js [hunk 6]
```diff
@@ -624,7 +624,7 @@ describe('$firebaseArray', function () {
 
     it('should apply $$defaults if they exist', function() {
       var arr = stubArray(STUB_DATA, $firebaseArray.$extend({
-        $$defaults: {aString: 'not_applied', foo: 'foo'}
+        $$defaults: {aString: 'alpha', foo: 'foo'}
```
Comment (katowulf) [inline tests/unit/FirebaseArray.spec.js]: Also here.

**Comment by katowulf** [comment]
@paulohp huge thanks for the PR and for appeasing the lawyers! A couple comments for you and we should be good to merge.

**Comment by paulohp** [comment]
@katowulf oh, I see it. I forget to fix it.. doing. thanks!

File: tests/unit/FirebaseObject.spec.js [hunk 7]
```diff
@@ -480,13 +480,6 @@ describe('$firebaseObject', function() {
       expect(obj.$remove()).toBeAPromise();
     });
 
-    it('should set $value to null and remove any local keys', function() {
```
Comment (jwngr) [inline tests/unit/FirebaseObject.spec.js]: I'm not sure I agree that this test should be removed. I'm fine getting rid of `$utils.dataKeys()`, but this test is still valid and should stick around.

**Comment by jwngr** [comment]
Looks like Travis CI didn't run for this PR. We should make sure to wait until Travis is fixed and passes before merging this in.

File: tests/unit/FirebaseObject.spec.js (hunk 7 above)
Comment (katowulf) [inline tests/unit/FirebaseObject.spec.js]: Great catch Jacob.

File: src/FirebaseArray.js (hunk 1 above)
Comment (jamestalmage) [inline src/FirebaseArray.js]: Oops, @katowulf is right on this one. This means `applyDefaults` needs to stay - it is slightly (but importantly) different than `angular.extend`. You could do `angular.extend(rec, this.$$defaults, rec)` but that's not as efficient as what we have, and a little confusing.

File: src/FirebaseAuth.js [hunk 8]
```diff
@@ -245,7 +245,7 @@
      * rejected if the client is unauthenticated and rejectIfAuthDataIsNull is true.
      */
     _routerMethodOnAuthPromise: function(rejectIfAuthDataIsNull) {
-      var ref = this._ref, utils = this._utils;
+      var ref = this._ref, utils = this._utils, q = this._q;
```
Comment (jwngr) [inline src/FirebaseAuth.js]: `utils` is defined but never used.

**Comment by jwngr** [comment]
Also make sure you run the test suite locally to ensure it all passes. Thanks!

File: src/FirebaseAuth.js (hunk 8 above)
Comment (paulohp) [inline src/FirebaseAuth.js]: thanks @jwngr.

**Comment by jamestalmage** [comment]
I just pulled it down and tried. The unit tests passed. That's not good. Removing `applyDefaults` should have broken things.

**Comment by paulohp** [comment]
@jwngr Yes, I'm running the `grunt test:unit` here all the time.

@jamestalmage I did some changes on unit tests and forgot to undo.. but I'm fixing it now.

**Comment by jamestalmage** [comment]
Ah, I see it's the `aString: 'not_applied'` ==> `aString:'alpha'` line @katowulf brought up. He's changed the expectation to make the test pass.

@paulohp as a general rule, you shouldn't change tests to make them pass. There can be exceptions to that rule if what you're changing has nothing to do with the public facing API (i.e. the underlying dependencies changed and you need to inject mocks a little differently, etc). But when you need to change the test inputs or the expected results, that's a huge red flag that you've just made a breaking change. This is the whole reason for unit tests. In this case they tell us that my idea to remove `applyDefaults` was a bad one.

Thanks for working on this PR (don't let my rant discourage you). It's awesome to see support from the community.

**Comment by paulohp** [comment]
@jamestalmage I know, thanks for your comment! Like I said above, I just modify the test to found where  the error comes from and to debug, but unfortunately I forgot to put things in place.

So, can I put the `applyDefaults` back in place and push the commit ? Maybe it's a good idea to put the test that @jwngr commented here: https://github.com/firebase/angularfire/pull/633#discussion_r33595974

Thanks!

**Comment by jamestalmage** [comment]
Yes, put `applyDefaults` back, and revert the discussed changes to the tests and push another commit.

A collaborator may ask you to squash your commits down the road, but don't worry about that until everyone has signed off on the changes.

**Comment by katowulf** [comment]
@paulohp status of this PR? All comments addressed? I only see a PR to address the .applyDefaults() and the missing test. Should there be others?

**Comment by paulohp** [comment]
@katowulf I'll finish this now. I'm trying to remember things that I need to do. Just `utils.batch` and `utils.wait`, right ?

cc @jamestalmage

**Comment by jamestalmage** [comment]
Oh man, I don't remember.

Just ping us when you think you have completed your changes. We will review and assess from there.

**Comment by katowulf** [comment]
Closing this for now. It can be reopened if the remaining issues are worked out.
